### ISSUE TITLE
no "charset_map" between the charsets "iso-8859-1" and "UTF-8"

### ISSUE THREAD

**Issue by locksmithunit** [engintron/engintron#1326]
2021/06/01 11:10:09 [error] 7384#7384: *9542 no "charset_map" between the charsets "iso-8859-1" and "UTF-8" while reading response header from upstream, client: ::ffff:66.249.66.133, server: localhost, request: "GET /sitemap_index.xml HTTP/1.1", subrequest: "/sitemap_index.xml", upstream: "http://23.235.206.148:8080/sitemap_index.xml", host: "locksmithunit.cat"

66.249.66.133 is a google crawler.
I think the error accrue because they tried to get in on IPv6 and I don't have IPv6,
only IPv4.

when you add cherset_map on nginx.conf the error is gone.
but i not sure this charset_map fix, will do the right job.

any recommendation?

**Comment by itsTheFae** [comment]
First off, nginx will represent IPv4 as an IPv6 mapped address with the configuration/listen directives as they are.  This can be a little confusing for those not used to IPv6 or dual-stack servers.  When you see `::ffff:` as a prefix to an IPv4 address, the request to nginx was most likely made via IPv4. 

The rest of this issue, regarding the charset_map is something relatively new.  It might be a bug in nginx, or something to do with the google crawler specifically.   From what I have been able to find out about it, it is likely an issue where PHP outputs content with one encoding, apache is given a content-type header that maybe doesn't match the encoding of the content, and then nginx charset module makes an assumption based on the header value it got from apache but can't operate on the content since it is encoded differently from the given header value. 
I haven't verified what exactly is going on here.  It doesn't seem to be a big issue for content in the browser though. So far Google tools haven't yelled at me about errors in my robots or sitemap files so I haven't really had the drive to figure it out. 

The things I've tried have not worked so far, so I don't have any recommendations.

**Comment by locksmithunit** [comment]
Ok i think you was right actually :)

So. in Whm have:
 WHM»Account Functions »Web Template Editor

now I see the templet from the hosting provider have  "iso-8859-1"
WHEN YOU CLICK DEFAULT
this changed to utf-8

i think this the isiue. and maybe worse, because this templet actually was in use in your system...
so I don't know what is more infected by this  "iso-8859-1".

but for everyone struggling with this isiue... is the hosting design the editor templet with stoped unicode

**Comment by locksmithunit** [comment]
Nah I still get the error even after fixing the templet.

**Comment by fevangelou** [comment]
The issue is most likely the XML itself. If you have control over it, ensure it defines UTF-8 for encoding...

**Comment by locksmithunit** [comment]
How can i check it? If you mean to the templet of the server, i change it already.
but still, I get the error :(

@fevangelou if you instruct me on how to check it I will be able to figure this out.

**Comment by locksmithunit** [comment]
if you mean to the sitemap-xml i get it on the robot.txt too and the other pages...
have a plugin control them (Yoast)
SURE YOU KNOW HIM, THIS WHAT CONTROLS ROBOTS.TXT, HTACSSES, AND THE SITE MAP.

**Comment by locksmithunit** [comment]
Yeah, i check it, WordPress (Yoast SEO plugin) controls that.
but still. you can change it and override that from APACHE (I don't know how to do it)

and have an option to fix it from Engintron, is will stop the error in the log,
but I don't know if is actually the proper way:
    
    ```
## Charset ##
    #charset                                  utf-8;
    #override_charset                       on;
    charset_map      iso-8859-1 utf-8 { }
```

@fevangelou what are your recommendations?
(if you know to explain how to fix it from cPanel, tell me I will do it)

**Comment by locksmithunit** [comment]
look at this @fevangelou :
http://nginx.org/en/docs/http/ngx_http_charset_module.html#charset_map

It can be mod secure as well with the PHP-FPM.
(I am not sure, I suspect modsecure change the charset... I talk about that with cPanel support)

BUT, IS NOT SO IMPORTANT, BECAUSE!
I think Nginx is able to override and fix this issue for all the clients of Engintron. (I see a lot of people with the same issue).
Nginx not the reason for this error, but he able to maybe override the utf-8

I only want to be sure is the right way to put it in nginx.conf
or you have other recommendations?

I think it can be fixed with X-Accel-Charset throw Nginx.
(this what they said in the link from nginx.com).

**Comment by locksmithunit** [comment]
Ok, after a lot of tests, me and cPanel support get it.
the issue is because Nginx took HTTP in the static, and Nginx design for UTF-8 read it from the PHP.

before 1 year something like that nginx must of it was get the HTTPS and APACHE was handling the HTTP.
no, because actually, APACHE gets out from the picture and Nginx / Engintron Hendel both, HTTP and https.
this is mean you need to convert the charset.
at least the Dynamic caching (HTML).

more than that.
server send header for English... (language and encoding problem):
https://www.w3.org/International/tutorials/tutorial-char-enc/

the issue is on ALL THE CPANEL.
@fevangelou 
I don't gonna lie, your plugin not causing it, and if he did, is not the fault of Engintron.

BUT!

Engintron can fix it easily by converting the encoding + language header for multiview.

if you provide me the right code and the right header, I will be able to fix it.
and I recommended fixing it for all the clients. (cPanel clients, this issues on ALL THE CPANEL)

Pull good update, an update can convert ISO to UTF:

```
## Charset ##
    #charset                                  utf-8;
    #override_charset                       on;
    charset_map      iso-8859-1 utf-8 { }
```

AND ACCEPT-LANGUAGE HEADER.

Sorry for all the messages, is because I talking with cPanel support directly, and anytime we check something also.

**Comment by locksmithunit** [comment]
OK... So i finish now with cPanel support and they blame Engintron...
the hole response goes from Engintron.

we even added charset UTF-8 in Apache MAIN INCULDED.
is ganarated more errors in the logs:
`2021/06/15 01:07:31 [error] 6271#6271: *4786 no "charset_map" between the charsets "iso-8859-1" and "UTF-8" while reading response header from upstream, client: ::ffff:66.249.65.225, server: localhost, request: "GET /sitemap_index.xml HTTP/1.1", subrequest: "/sitemap_index.xml", upstream: "http://23.235.206.148:8080/sitemap_index.xml", host: "locksmithunit.cat"`


they said is the /etc/nginx/mime.types of engintron.

you can fix it 2 ways... or in nginx.conf with charset utf-8;  ( but need the right charset_map for iso-8859-1 utf-8 { } )

or fix the MIME.TYPES of engintron... (this what cPanel support said)
btw: 66.249.65.225
is Googlebot... proxy google bot.

sites not actually read the SEO properly...
Please help @fevangelou

**Comment by locksmithunit** [comment]
is comes from engintron.
http_common

i checked another nginx similar to yours, usually, XML and Robots.txt is not a good idea.
when I bypass inside the http_common the robots.txt, I stop to get the error.
does the mean error come from the cache?

another thing is Unicode, have a problem with encoding. i figure out is not only google, is aherf and a lot of important bots:

2021/06/16 08:29:33 [error] 23726#23726: *286 locksmithunit.cat could not be resolved (2: Server failure), client: ::ffff:54.155.164.35, server: localhost, request: "GET / HTTP/1.1", host: "locksmithunit.cat"

54.155.164.35 - aherf SEO backlinks IP
this happened when I bypass the cache inside robots.txt throw Engintron.
btw the robots.txt gets hit only when I remove the location block of robots.txt completely.

I think I better remove it, maybe the favicon as well...
give to apache to  Hendel, they maybe will fix the charset error in the robots.txt

but is 100% comes from engintron HTTP cache.

**Comment by locksmithunit** [comment]
issue not from engintron, is solved, you was right, after I reinstall the Engintron and delete him several times is got ok,
I suspect on the log configuration so I open topic about that and close this one.
